README.md
# Bevy Tweening

THIS IS A FORKED REPO OF https://github.com/djeedai/bevy_extra/tree/main/crates/bevy_tweening, I DID NOT WRITE THE ORIGINAL CODE FOR THIS, just made some edits
to allow auto removal of components if they are animating once and done.

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) [![Doc](https://docs.rs/bevy_tweening/badge.svg)](https://docs.rs/bevy_tweening) [![Crate](https://img.shields.io/crates/v/bevy_tweening.svg)](https://crates.io/crates/bevy_tweening)

Tweening animation plugin for Bevy.

## Usage

### System setup

Add the tweening plugin to your app:

```rust
App::default()
    .add_default_plugins()
    .add_plugin(TweeningPlugin)
    .run();
```

### Animate a component

Animate the transform position of an entity:

```rust
commands
    // Spawn a Sprite entity to animate the position of
    .spawn_bundle(SpriteBundle {
        sprite: Sprite {
            color: Color::RED,
            custom_size: Some(Vec2::new(size, size)),
            ..Default::default()
        },
        ..Default::default()
    })
    // Add an Animator component to perform the animation
    .insert(Animator::new(
        // Use a quadratic easing on both endpoints
        EaseFunction::QuadraticInOut,
        // Loop animation back and forth over 1 second, with a 0.5 second
        // pause after each cycle (start -> end -> start).
        TweeningType::PingPong {
            duration: Duration::from_secs(1),
            pause: Some(Duration::from_millis(500)),
        },
        // The lens gives access to the Transform component of the Sprite,
        // for the Animator to animate it. It also contains the start and
        // end values associated with the animation ratios 0. and 1.
        TransformPositionLens {
            start: Vec3::new(0., 0., 0.),
            end: Vec3::new(1., 2., -4.),
        },
    ));
```

## Predefined Lenses

The naming scheme for predefined lenses is `"<TargetName><FieldName>Lens"`, where `<TargetName>` is the name of the target component or asset which is queried, and `<FieldName>` is the field which is mutated in place.

### Bevy Components

| Target Component | Animated Field | Lens |
|---|---|---|
| [`Sprite`](https://docs.rs/bevy/0.6.0/bevy/sprite/struct.Sprite.html) | [`color`](https://docs.rs/bevy/0.6.0/bevy/sprite/struct.Sprite.html#structfield.color) | [`SpriteColorLens`](https://docs.rs/bevy_tweening/latest/bevy_tweening/struct.SpriteColorLens.html) |
| [`Transform`](https://docs.rs/bevy/0.6.0/bevy/transform/components/struct.Transform.html) | [`translation`](https://docs.rs/bevy/0.6.0/bevy/transform/components/struct.Transform.html#structfield.translation) | [`TransformPositionLens`](https://docs.rs/bevy_tweening/latest/bevy_tweening/struct.TransformPositionLens.html) |
| | [`rotation`](https://docs.rs/bevy/0.6.0/bevy/transform/components/struct.Transform.html#structfield.rotation) | [`TransformRotationLens`](https://docs.rs/bevy_tweening/latest/bevy_tweening/struct.TransformRotationLens.html) |
| | [`scale`](https://docs.rs/bevy/0.6.0/bevy/transform/components/struct.Transform.html#structfield.scale) | [`TransformScaleLens`](https://docs.rs/bevy_tweening/latest/bevy_tweening/struct.TransformScaleLens.html) |
| [`Style`](https://docs.rs/bevy/0.6.0/bevy/ui/struct.Style.html) | [`position`](https://docs.rs/bevy/0.6.0/bevy/ui/struct.Style.html#structfield.position) | [`UiPositionLens`](https://docs.rs/bevy_tweening/latest/bevy_tweening/struct.UiPositionLens.html) |
| [`Text`](https://docs.rs/bevy/0.6.0/bevy/text/struct.Text.html) | [`TextStyle::color`](https://docs.rs/bevy/0.6.0/bevy/text/struct.TextStyle.html#structfield.color) | [`TextColorLens`](https://docs.rs/bevy_tweening/latest/bevy_tweening/struct.TextColorLens.html) |

### Bevy Assets

| Target Asset | Animated Field | Lens |
|---|---|---|
| [`ColorMaterial`](https://docs.rs/bevy/0.6.0/bevy/sprite/struct.ColorMaterial.html) | [`color`](https://docs.rs/bevy/0.6.0/bevy/sprite/struct.ColorMaterial.html#structfield.color) | [`ColorMaterialColorLens`](https://docs.rs/bevy_tweening/latest/bevy_tweening/struct.ColorMaterialColorLens.html) |

### Custom component support

To be able to animate some fields of a custom component, a custom lens need to be implemented for that component, which **linearly** interpolates the field(s) of that component.

```rust
#[derive(Component)]
struct CustomComponent(f32);

struct CustomLens {
    start: f32,
    end: f32,
}

impl Lens<CustomComponent> for CustomLens {
    fn lerp(&self, target: &mut CustomComponent, ratio: f32) -> f32 {
        target.0 = self.start + (self.end - self.start) * ratio;
    }
}
```

This process can also be used to interpolate fields of existing Bevy built-in components for which a predfined lens is not provided.

The basic formula for lerp (linear interpolation) is either of:

- `start + (end - start) * scalar`
- `start * (1.0 - scalar) + end * scalar`

The two formulations are mathematically equivalent, but one may be more suited than the other depending on the type interpolated and the operations available, and the potential floating-point precision errors.

Then, the system `component_animator_system::<CustomComponent>` needs to be added to the application.

## Custom asset support

The process is similar to custom components, creating a custom lens for the custom asset. The system to add is `asset_animator_system::<CustomAsset>`.

## Examples

See the [`examples/`](https://github.com/djeedai/bevy_extra/tree/main/crates/bevy_tweening/examples) folder.

### [`sprite_color`](examples/sprite_color.rs)

```rust
cargo run --example sprite_color --features="bevy/bevy_winit"
```

![sprite_color](https://raw.githubusercontent.com/djeedai/bevy_extra/main/crates/bevy_tweening/examples/sprite_color.gif)

### [`transform_rotation`](examples/transform_rotation.rs)

```rust
cargo run --example transform_rotation --features="bevy/bevy_winit"
```

![sprite_color](https://raw.githubusercontent.com/djeedai/bevy_extra/main/crates/bevy_tweening/examples/transform_rotation.gif)

### [`transform_translation`](examples/transform_translation.rs)

```rust
cargo run --example transform_translation --features="bevy/bevy_winit"
```

![sprite_color](https://raw.githubusercontent.com/djeedai/bevy_extra/main/crates/bevy_tweening/examples/transform_translation.gif)

### [`colormaterial_color`](examples/colormaterial_color.rs)

```rust
cargo run --example colormaterial_color --features="bevy/bevy_winit"
```

![colormaterial_color](https://raw.githubusercontent.com/djeedai/bevy_extra/main/crates/bevy_tweening/examples/colormaterial_color.gif)

## Ease Functions

Many [ease functions](https://docs.rs/interpolation/0.2.0/interpolation/enum.EaseFunction.html) are available:

- QuadraticIn
- QuadraticOut
- QuadraticInOut
- CubicIn
- CubicOut
- CubicInOut
- QuarticIn
- QuarticOut
- QuarticInOut
- QuinticIn
- QuinticOut
- QuinticInOut
- SineIn
- SineOut
- SineInOut
- CircularIn
- CircularOut
- CircularInOut
- ExponentialIn
- ExponentialOut
- ExponentialInOut
- ElasticIn
- ElasticOut
- ElasticInOut
- BackIn
- BackOut
- BackInOut
- BounceIn
- BounceOut
- BounceInOut

## Comparison with `bevy_easings`

The `bevy_tweening` library started as a fork of [the `bevy_easings` library by François Mocker](https://github.com/mockersf/bevy_extra/blob/master/bevy_easings), with the goals to:

- explore an alternative design based on lenses instead of generic types for each easer/animator. This reduces both the number of generic types needed, and hopefully the code size, as well as the number of systems needed to perform the interpolation.
- improve the interpolation of assets to avoid creating many copies like `bevy_easings` does, and instead mutate the assets (and, by similarity, the components too) in-place without making a copy. The in-place mutation also allows a more optimal interpolation limited to modifying the fields of interest only, instead of creating a new copy of the entire component each tick.


CHANGELOG
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2022-01-09

### Changed

- Update to Bevy 0.6
- Update to rust edition 2021
- Force Cargo resolver v2

### Added

- Added built-in lens `SpriteColorLens`, since the color of a `Sprite` is now an intrinsic property of the component in Bevy 0.6, and does not use `ColorMaterial` anymore.

## [0.1.0] - 2021-12-24

Initial version for Bevy 0.5


src/plugin.rs
use bevy::{asset::Asset, ecs::component::Component, prelude::*};

use crate::{
    Animate, Animator, AnimatorState, AssetAnimator, Lens, ScaleAnimator, TransformPositionLens,
    TransformScaleLens, TranslationAnimator, TweeningType,
};

/// Plugin to add systems related to tweening
#[derive(Debug, Clone, Copy)]
pub struct TweeningPlugin;

impl Plugin for TweeningPlugin {
    fn build(&self, app: &mut App) {
        app.add_system(component_animator_system::<Transform>)
            .add_system(component_animator_system::<Text>)
            .add_system(component_animator_system::<Style>)
            .add_system(component_animator_system::<Sprite>)
            .add_system_set_to_stage(
                CoreStage::PreUpdate,
                SystemSet::new()
                    .with_system(animate_animator_system::<ScaleAnimator, TransformScaleLens, Transform>)
                    .with_system(animate_animator_system::<TranslationAnimator, TransformPositionLens, Transform>)
            )
            .add_system(asset_animator_system::<ColorMaterial>);
    }
}

pub fn animate_animator_system<T, U, V>(
    mut commands: Commands,
    time: Res<Time>,
    mut query: Query<(Entity, &mut V, &mut T)>,
) where
    T: Animate<U, V> + Component,
    U: Lens<V>,
    V: Component,
{
    for (entity, ref mut target, ref mut animator) in query.iter_mut() {
        if animator.get_state() == AnimatorState::Playing {
            animator.get_timer().tick(time.delta());
        }
        if *animator.is_paused() {
            if animator.get_timer().just_finished() {
                match animator.get_tweening_type() {
                    TweeningType::Once { duration } => {
                        animator.get_timer().set_duration(duration);
                    }
                    TweeningType::Loop { duration, .. } => {
                        animator.get_timer().set_duration(duration);
                    }
                    TweeningType::PingPong { duration, .. } => {
                        animator.get_timer().set_duration(duration);
                    }
                }
                animator.get_timer().reset();
                *animator.is_paused() = false;
            }
        } else {
            if animator.get_timer().duration().as_secs_f32() != 0. {
                let progress = if animator.get_direction().is_positive() {
                    animator.get_timer().percent()
                } else {
                    animator.get_timer().percent_left()
                };
                let factor = animator.get_easing_function().sample(progress);
                animator.apply(target.as_mut(), factor);
            }
            if animator.get_timer().finished() {
                match animator.get_tweening_type() {
                    TweeningType::Once { .. } => {
                        commands.entity(entity).remove::<T>();
                    }
                    TweeningType::Loop { pause, .. } => {
                        if let Some(pause) = pause {
                            animator.get_timer().set_duration(pause);
                            *animator.is_paused() = true;
                        }
                        animator.get_timer().reset();
                    }
                    TweeningType::PingPong { pause, .. } => {
                        if let Some(pause) = pause {
                            animator.get_timer().set_duration(pause);
                            *animator.is_paused() = true;
                        }
                        animator.get_timer().reset();
                        *animator.get_direction() *= -1;
                    }
                }
            }
        }
    }
}

pub fn component_animator_system<T: Component>(
    mut commands: Commands,
    time: Res<Time>,
    mut query: Query<(Entity, &mut T, &mut Animator<T>)>,
) {
    for (entity, ref mut target, ref mut animator) in query.iter_mut() {
        if animator.state == AnimatorState::Playing {
            animator.timer.tick(time.delta());
        }
        if animator.paused {
            if animator.timer.just_finished() {
                match animator.tweening_type {
                    TweeningType::Once { duration } => {
                        animator.timer.set_duration(duration);
                    }
                    TweeningType::Loop { duration, .. } => {
                        animator.timer.set_duration(duration);
                    }
                    TweeningType::PingPong { duration, .. } => {
                        animator.timer.set_duration(duration);
                    }
                }
                animator.timer.reset();
                animator.paused = false;
            }
        } else {
            if animator.timer.duration().as_secs_f32() != 0. {
                let progress = if animator.direction.is_positive() {
                    animator.timer.percent()
                } else {
                    animator.timer.percent_left()
                };
                let factor = animator.ease_function.sample(progress);
                animator.apply(target, factor);
            }
            if animator.timer.finished() {
                match animator.tweening_type {
                    TweeningType::Once { .. } => {
                        commands.entity(entity).remove::<Animator<T>>();
                    }
                    TweeningType::Loop { pause, .. } => {
                        if let Some(pause) = pause {
                            animator.timer.set_duration(pause);
                            animator.paused = true;
                        }
                        animator.timer.reset();
                    }
                    TweeningType::PingPong { pause, .. } => {
                        if let Some(pause) = pause {
                            animator.timer.set_duration(pause);
                            animator.paused = true;
                        }
                        animator.timer.reset();
                        animator.direction *= -1;
                    }
                }
            }
        }
    }
}

pub fn asset_animator_system<T: Asset>(
    time: Res<Time>,
    mut assets: ResMut<Assets<T>>,
    mut query: Query<&mut AssetAnimator<T>>,
) {
    for ref mut animator in query.iter_mut() {
        if animator.state == AnimatorState::Playing {
            animator.timer.tick(time.delta());
        }
        if animator.paused {
            if animator.timer.just_finished() {
                match animator.tweening_type {
                    TweeningType::Once { duration } => {
                        animator.timer.set_duration(duration);
                    }
                    TweeningType::Loop { duration, .. } => {
                        animator.timer.set_duration(duration);
                    }
                    TweeningType::PingPong { duration, .. } => {
                        animator.timer.set_duration(duration);
                    }
                }
                animator.timer.reset();
                animator.paused = false;
            }
        } else {
            if animator.timer.duration().as_secs_f32() != 0. {
                let progress = if animator.direction.is_positive() {
                    animator.timer.percent()
                } else {
                    animator.timer.percent_left()
                };
                let factor = animator.ease_function.sample(progress);
                if let Some(target) = assets.get_mut(animator.handle()) {
                    animator.apply(target, factor);
                }
            }
            if animator.timer.finished() {
                match animator.tweening_type {
                    TweeningType::Once { .. } => {
                        // commands.entity(entity).remove::<Animator>();
                    }
                    TweeningType::Loop { pause, .. } => {
                        if let Some(pause) = pause {
                            animator.timer.set_duration(pause);
                            animator.paused = true;
                        }
                        animator.timer.reset();
                    }
                    TweeningType::PingPong { pause, .. } => {
                        if let Some(pause) = pause {
                            animator.timer.set_duration(pause);
                            animator.paused = true;
                        }
                        animator.timer.reset();
                        animator.direction *= -1;
                    }
                }
            }
        }
    }
}


benchmarks/benches/lens.rs
#[macro_use]
extern crate criterion;

use bevy::prelude::*;
use bevy_tweening::*;
use criterion::{black_box, Criterion};

fn text_color_lens(c: &mut Criterion) {
    let mut lens = TextColorLens {
        start: Color::RED,
        end: Color::BLUE,
        section: 0,
    };
    let mut text = Text::with_section(
        "test".to_string(),
        TextStyle {
            font: Default::default(),
            font_size: 60.0,
            color: Color::WHITE,
        },
        TextAlignment {
            vertical: VerticalAlign::Center,
            horizontal: HorizontalAlign::Center,
        },
    );
    c.bench_function("TextColorLens", |b| {
        b.iter(|| lens.interpolate(&mut text, black_box(0.3)))
    });
}

fn transform_position_lens(c: &mut Criterion) {
    let mut lens = TransformPositionLens {
        start: Vec3::ZERO,
        end: Vec3::ONE,
    };
    let mut transform = Transform::identity();
    c.bench_function("TransformPositionLens", |b| {
        b.iter(|| lens.interpolate(&mut transform, black_box(0.3)))
    });
}

fn transform_rotation_lens(c: &mut Criterion) {
    let mut lens = TransformRotationLens {
        start: Quat::IDENTITY,
        end: Quat::from_rotation_x(72.0_f32.to_radians()),
    };
    let mut transform = Transform::identity();
    c.bench_function("TransformRotationLens", |b| {
        b.iter(|| lens.interpolate(&mut transform, black_box(0.3)))
    });
}

fn transform_scale_lens(c: &mut Criterion) {
    let mut lens = TransformScaleLens {
        start: Vec3::ONE,
        end: Vec3::new(1.5, 2.0, 3.0),
    };
    let mut transform = Transform::identity();
    c.bench_function("TransformScaleLens", |b| {
        b.iter(|| lens.interpolate(&mut transform, black_box(0.3)))
    });
}

criterion_group!(
    benches,
    text_color_lens,
    transform_position_lens,
    transform_rotation_lens,
    transform_scale_lens
);
criterion_main!(benches);


src/lib.rs
#![deny(
    warnings,
    missing_copy_implementations,
    trivial_casts,
    trivial_numeric_casts,
    unsafe_code,
    unstable_features,
    unused_import_braces,
    unused_qualifications,
    missing_docs
)]

//! Tweening animation plugin for the Bevy game engine
//!
//! This library provides interpolation-based animation between ("tweening") two values, for a variety
//! of Bevy components and assets. Each field of a component or asset can be animated via a collection
//! or predefined easing functions, or providing a custom animation curve.
//!
//! # Example
//!
//! Add the tweening plugin to your app:
//!
//! ```rust
//! # use bevy::prelude::*;
//! # use bevy_tweening::*;
//! App::default()
//!     .add_default_plugins()
//!     .add_plugin(TweeningPlugin)
//!     .run();
//! ```
//!
//! Animate the position ([`Transform::translation`]) of an [`Entity`]:
//!
//! ```rust
//! # use bevy_tweening::*;
//! # use std::time::Duration;
//! commands
//!     // Spawn a Sprite entity to animate the position of
//!     .spawn_bundle(SpriteBundle {
//!         sprite: Sprite {
//!             color: Color::RED,
//!             custom_size: Some(Vec2::new(size, size)),
//!             ..Default::default()
//!         },
//!         ..Default::default()
//!     })
//!     // Add an Animator component to perform the animation
//!     .insert(Animator::new(
//!         // Use a quadratic easing on both endpoints
//!         EaseFunction::QuadraticInOut,
//!         // Loop animation back and forth over 1 second, with a 0.5 second
//!         // pause after each cycle (start -> end -> start).
//!         TweeningType::PingPong {
//!             duration: Duration::from_secs(1),
//!             pause: Some(Duration::from_millis(500)),
//!         },
//!         // The lens gives access to the Transform component of the Sprite,
//!         // for the Animator to animate it. It also contains the start and
//!         // end values associated with the animation ratios 0. and 1.
//!         TransformPositionLens {
//!             start: Vec3::new(0., 0., 0.),
//!             end: Vec3::new(1., 2., -4.),
//!         },
//!     ));
//! ```
//!
//! # Animators and lenses
//!
//! Bevy components and assets are animated with tweening animator components. Those animators determine
//! the fields to animate using lenses.
//!
//! ## Components animation
//!
//! Components are animated with the [`Animator`] component, which is generic over the type of component
//! it animates. This is a restriction imposed by Bevy, to access the animated component as a mutable
//! reference via a [`Query`] and comply with the ECS rules.
//!
//! The [`Animator`] itself is not generic over the subset of fields of the components it animates.
//! This limits the proliferation of generic types when animating e.g. both the position and rotation
//! of an entity.
//!
//! ## Assets animation
//!
//! Assets are animated in a similar way to component, via the [`AssetAnimator`] component. Because assets
//! are typically shared, and the animation applies to the asset itself, all users of the asset see the
//! animation. For example, animating the color of a [`ColorMaterial`] will change the color of all the
//! 2D meshes using that material.
//!
//! ## Lenses
//!
//! Both [`Animator`] and [`AssetAnimator`] access the field(s) to animate via a lens, a type that implements
//! the [`Lens`] trait. Several predefined lenses are provided for the most commonly animated fields, like the
//! components of a [`Transform`]. A custom lens can also be created by implementing the trait, allowing to
//! animate virtually any field of any Bevy component or asset.
//!
//! [`Transform::translation`]: https://docs.rs/bevy/0.6.0/bevy/transform/components/struct.Transform.html#structfield.translation
//! [`Entity`]: https://docs.rs/bevy/0.6.0/bevy/ecs/entity/struct.Entity.html
//! [`Query`]: https://docs.rs/bevy/0.6.0/bevy/ecs/system/struct.Query.html
//! [`ColorMaterial`]: https://docs.rs/bevy/0.6.0/bevy/sprite/struct.ColorMaterial.html
//! [`Sprite`]: https://docs.rs/bevy/0.6.0/bevy/sprite/struct.Sprite.html
//! [`Transform`]: https://docs.rs/bevy/0.6.0/bevy/transform/components/struct.Transform.html

use std::time::Duration;

use bevy::{asset::Asset, prelude::*};

use interpolation::Ease as IEase;
pub use interpolation::EaseFunction;
pub use interpolation::Lerp;

mod lens;
mod plugin;

pub use lens::{
    ColorMaterialColorLens, Lens, SpriteColorLens, TextColorLens, TransformPositionLens,
    TransformRotationLens, TransformScaleLens, UiPositionLens,
};
pub use plugin::TweeningPlugin;

/// How should this easing loop repeat
#[derive(Clone, Copy)]
pub enum TweeningType {
    /// Only happen once
    Once {
        /// duration of the easing
        duration: Duration,
    },
    /// Looping, restarting from the start once finished
    Loop {
        /// duration of the easing
        duration: Duration,
        /// duration of the pause between two loops
        pause: Option<Duration>,
    },
    /// Repeat the animation back and forth
    PingPong {
        /// duration of the easing
        duration: Duration,
        /// duration of the pause before starting again in the other direction
        pause: Option<Duration>,
    },
}

/// Playback state of an animator.
#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum AnimatorState {
    /// The animation is playing.
    Playing,
    /// The animation is paused/stopped.
    Paused,
}

impl std::ops::Not for AnimatorState {
    type Output = AnimatorState;

    fn not(self) -> Self::Output {
        match self {
            AnimatorState::Paused => AnimatorState::Playing,
            AnimatorState::Playing => AnimatorState::Paused,
        }
    }
}

/// Describe how eased value should be computed.
#[derive(Clone, Copy)]
pub enum EaseMethod {
    /// Follow `EaseFunction`.
    EaseFunction(EaseFunction),
    /// Linear interpolation, with no function.
    Linear,
    /// Discrete interpolation, eased value will jump from start to end when
    /// stepping over the discrete limit.
    Discrete(f32),
    /// Use a custom function to interpolate the value.
    CustomFunction(fn(f32) -> f32),
}

impl EaseMethod {
    fn sample(self, x: f32) -> f32 {
        match self {
            EaseMethod::EaseFunction(function) => x.calc(function),
            EaseMethod::Linear => x,
            EaseMethod::Discrete(limit) => {
                if x > limit {
                    1.
                } else {
                    0.
                }
            }
            EaseMethod::CustomFunction(function) => function(x),
        }
    }
}

impl Into<EaseMethod> for EaseFunction {
    fn into(self) -> EaseMethod {
        EaseMethod::EaseFunction(self)
    }
}

/// trait for animating
pub trait Animate<T, U>
where
    T: Lens<U>,
    U: Component,
{
    // /// The type of the lens
    // type Lens;

    /// apply interpolation to the component
    fn apply(&mut self, target: &mut U, ratio: f32);

    /// get the tweening type of the animation
    fn get_tweening_type(&self) -> TweeningType;

    /// get the easing function
    fn get_easing_function(&self) -> EaseMethod;

    /// get animation state
    fn get_state(&self) -> AnimatorState;

    /// get the animation timing
    fn get_timer(&mut self) -> &mut Timer;

    /// checks if animation is paused
    fn is_paused(&mut self) -> &mut bool;

    /// get direction of the animation
    fn get_lens(&self) -> T;

    /// get animation direction
    fn get_direction(&mut self) -> &mut i16;
}

/// Component to control the animation of another component, with translation
#[derive(Component)]
pub struct ScaleAnimator {
    ease_function: EaseMethod,
    timer: Timer,

    /// Control if this animation is played or not.
    pub state: AnimatorState,
    paused: bool,
    tweening_type: TweeningType,
    direction: i16,
    lens: TransformScaleLens,
}

impl std::fmt::Debug for ScaleAnimator {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Animator")
            .field("state", &self.state)
            .finish()
    }
}

impl ScaleAnimator {
    /// Create a new animator component from an easing function, tweening type, and a lens.
    /// The type `T` of the component to animate can generally be deducted from the lens type itself.
    pub fn new(
        ease_function: impl Into<EaseMethod>,
        tweening_type: TweeningType,
        lens: TransformScaleLens,
    ) -> Self {
        ScaleAnimator {
            ease_function: ease_function.into(),
            timer: match tweening_type {
                TweeningType::Once { duration } => Timer::new(duration, false),
                TweeningType::Loop { duration, .. } => Timer::new(duration, false),
                TweeningType::PingPong { duration, .. } => Timer::new(duration, false),
            },
            state: AnimatorState::Playing,
            paused: false,
            tweening_type,
            direction: 1,
            lens: lens,
        }
    }
}

impl Animate<TransformScaleLens, Transform> for ScaleAnimator {
    #[inline(always)]
    fn apply(&mut self, target: &mut Transform, ratio: f32) {
        self.lens.lerp(target, ratio);
    }

    fn get_tweening_type(&self) -> TweeningType {
        self.tweening_type
    }

    fn get_easing_function(&self) -> EaseMethod {
        self.ease_function
    }

    fn get_state(&self) -> AnimatorState {
        self.state
    }

    fn get_timer(&mut self) -> &mut Timer {
        &mut self.timer
    }

    fn is_paused(&mut self) -> &mut bool {
        &mut self.paused
    }

    fn get_lens(&self) -> TransformScaleLens {
        self.lens
    }

    fn get_direction(&mut self) -> &mut i16 {
        &mut self.direction
    }
}

/// Component to control the animation of another component, with translation
#[derive(Component)]
pub struct TranslationAnimator {
    ease_function: EaseMethod,
    timer: Timer,

    /// Control if this animation is played or not.
    pub state: AnimatorState,
    paused: bool,
    tweening_type: TweeningType,
    direction: i16,
    lens: TransformPositionLens,
}

impl std::fmt::Debug for TranslationAnimator {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Animator")
            .field("state", &self.state)
            .finish()
    }
}

impl TranslationAnimator {
    /// Create a new animator component from an easing function, tweening type, and a lens.
    /// The type `T` of the component to animate can generally be deducted from the lens type itself.
    pub fn new(
        ease_function: impl Into<EaseMethod>,
        tweening_type: TweeningType,
        lens: TransformPositionLens,
    ) -> Self {
        TranslationAnimator {
            ease_function: ease_function.into(),
            timer: match tweening_type {
                TweeningType::Once { duration } => Timer::new(duration, false),
                TweeningType::Loop { duration, .. } => Timer::new(duration, false),
                TweeningType::PingPong { duration, .. } => Timer::new(duration, false),
            },
            state: AnimatorState::Playing,
            paused: false,
            tweening_type,
            direction: 1,
            lens: lens,
        }
    }
}

impl Animate<TransformPositionLens, Transform> for TranslationAnimator {
    #[inline(always)]
    fn apply(&mut self, target: &mut Transform, ratio: f32) {
        self.lens.lerp(target, ratio);
    }

    fn get_tweening_type(&self) -> TweeningType {
        self.tweening_type
    }

    fn get_easing_function(&self) -> EaseMethod {
        self.ease_function
    }

    fn get_state(&self) -> AnimatorState {
        self.state
    }

    fn get_timer(&mut self) -> &mut Timer {
        &mut self.timer
    }

    fn is_paused(&mut self) -> &mut bool {
        &mut self.paused
    }

    fn get_lens(&self) -> TransformPositionLens {
        self.lens
    }

    fn get_direction(&mut self) -> &mut i16 {
        &mut self.direction
    }
}

/// Component to control the animation of another component.
#[derive(Component)]
pub struct Animator<T> {
    ease_function: EaseMethod,
    timer: Timer,
    /// Control if this animation is played or not.
    pub state: AnimatorState,
    paused: bool,
    tweening_type: TweeningType,
    direction: i16,
    lens: Box<dyn Lens<T> + Send + Sync + 'static>,
}

impl<T: std::fmt::Debug> std::fmt::Debug for Animator<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Animator")
            .field("state", &self.state)
            .finish()
    }
}

impl<T> Animator<T> {
    /// Create a new animator component from an easing function, tweening type, and a lens.
    /// The type `T` of the component to animate can generally be deducted from the lens type itself.
    pub fn new<L>(
        ease_function: impl Into<EaseMethod>,
        tweening_type: TweeningType,
        lens: L,
    ) -> Self
    where
        L: Lens<T> + Send + Sync + 'static,
    {
        Animator {
            ease_function: ease_function.into(),
            timer: match tweening_type {
                TweeningType::Once { duration } => Timer::new(duration, false),
                TweeningType::Loop { duration, .. } => Timer::new(duration, false),
                TweeningType::PingPong { duration, .. } => Timer::new(duration, false),
            },
            state: AnimatorState::Playing,
            paused: false,
            tweening_type,
            direction: 1,
            lens: Box::new(lens),
        }
    }

    #[inline(always)]
    fn apply(&mut self, target: &mut T, ratio: f32) {
        self.lens.lerp(target, ratio);
    }
}

/// Component to control the animation of an asset.
#[derive(Component)]
pub struct AssetAnimator<T: Asset> {
    ease_function: EaseMethod,
    timer: Timer,
    /// Control if this animation is played or not.
    pub state: AnimatorState,
    paused: bool,
    tweening_type: TweeningType,
    direction: i16,
    lens: Box<dyn Lens<T> + Send + Sync + 'static>,
    handle: Handle<T>,
}

impl<T: Asset + std::fmt::Debug> std::fmt::Debug for AssetAnimator<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("AssetAnimator")
            .field("state", &self.state)
            .finish()
    }
}

impl<T: Asset> AssetAnimator<T> {
    /// Create a new asset animator component from an easing function, tweening type, and a lens.
    /// The type `T` of the asset to animate can generally be deducted from the lens type itself.
    /// The component can be attached on any entity.
    pub fn new<L>(
        handle: Handle<T>,
        ease_function: impl Into<EaseMethod>,
        tweening_type: TweeningType,
        lens: L,
    ) -> Self
    where
        L: Lens<T> + Send + Sync + 'static,
    {
        AssetAnimator {
            ease_function: ease_function.into(),
            timer: match tweening_type {
                TweeningType::Once { duration } => Timer::new(duration, false),
                TweeningType::Loop { duration, .. } => Timer::new(duration, false),
                TweeningType::PingPong { duration, .. } => Timer::new(duration, false),
            },
            state: AnimatorState::Playing,
            paused: false,
            tweening_type,
            direction: 1,
            lens: Box::new(lens),
            handle,
        }
    }

    fn handle(&self) -> Handle<T> {
        self.handle.clone()
    }

    #[inline(always)]
    fn apply(&mut self, target: &mut T, ratio: f32) {
        self.lens.lerp(target, ratio);
    }
}
